modules/page-memory/src/main/java/org/apache/ignite/internal/pagememory/io/AbstractDataPageIo.java

```diff
@@ -1438,6 +1438,22 @@ protected abstract void writeRowData(
             boolean newRow
     ) throws IgniteInternalCheckedException;
 
+
+    /**
+     * Narrows an {@code int} down to {@code short} throwing an exception if the value cannot be exactly represented as a {@code short}.
+     *
+     * @param intValue Value to narrow down.
+     * @return Resulting short value.
+     * @throws IllegalArgumentException If the provided value does not fit the {@code short} range.
+     */
+    protected static short narrowIntToShort(int intValue) {
```

Comment:
Similar check is already in `org.apache.ignite.internal.pagememory.io.AbstractDataPageIo#addRow`, we can expand the existing one if necessary.

Reply:
If I'm not mistaken, the check in `AbstractDataPageIo#addRow()` is made with `assert` keyword, so it might absent in a production environment, hence it cannot be relied upon. I'd prefer to either leave the check that I added or change the check in `addRow()` from `assert` to throwing an exception.

WDYT?

Reply:
I think the assertion should be enough and check all this in tests. 
In production code, this check is not needed.

Reply:
Can a user configure a really large page, like 1Mb? If yes, then free space in a page might be around 1Mb, which is way more than 64kb. If the user does not have `-ea` (because they do this in production), they will get strange behavior instead of a proper exception.

Assertions should be used only for catching programmer errors (i.e. our errors). They should not be used to protect against a possible misconfiguration. Also, tests would not help in such scenario.

Reply:
Current limit is 32k I believe

Reply:
I do not think that users will use our product immediately in production without test runs on test environments with -ea.

Reply:
Removed the additional check. But this should not be forgotten later.